File: backend/Models/Course.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace DevEdu.Api.Models;

// ─── Lokalisierung ────────────────────────────────────────────────────────────
// BsonIgnoreExtraElements: toleriert importierte Dokumente mit zusätzlichen/
// abweichenden Feldern (z. B. ein literales `Id` statt `_id` in Sub-Dokumenten),
// statt beim Deserialisieren zu werfen.

[BsonIgnoreExtraElements]
public class TextItem
{
    public string Text { get; set; } = string.Empty;
    /// <summary>0 = Russian, 1 = German, 2 = English</summary>
    public int Language { get; set; }
}

[BsonIgnoreExtraElements]
public class Texte
{
    public List<TextItem> Items { get; set; } = new();
}

// ─── Enums ───────────────────────────────────────────────────────────────────

public enum ChapterContentType { Lesson = 0, Video = 1, Questions = 2 }

// ─── ChapterContent ──────────────────────────────────────────────────────────

[BsonIgnoreExtraElements]
public class ChapterContent
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string ElementId { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string CourseId { get; set; } = string.Empty;
    public string ChapterId { get; set; } = string.Empty;
    public Texte Titel { get; set; } = new();
    public ChapterContentType ContentType { get; set; } = ChapterContentType.Lesson;
    public string LessonText { get; set; } = string.Empty;
    public Texte LessonTexte { get; set; } = new();
    public string VideoUrl { get; set; } = string.Empty;
    /// <summary>ID des QuestionList-Dokuments (leer wenn kein Quiz).</summary>
    public string QuestionListId { get; set; } = string.Empty;
    public int SortOrder { get; set; }
    public double AverageRank { get; set; }
    public int MaxRank { get; set; }
}

// ─── Chapter ─────────────────────────────────────────────────────────────────

[BsonIgnoreExtraElements]
public class Chapter
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string ElementId { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string CourseId { get; set; } = string.Empty;
    public Texte Titel { get; set; } = new();
    public int SortOrder { get; set; }
    public bool Show { get; set; } = true;
    public List<ChapterContent> ChapterContent { get; set; } = new();

    // ─── F8: Kapitel-Abschlussquiz ───────────────────────────────────────────
    /// <summary>ID der QuestionList des Kapitel-Abschlussquiz (null = kein Quiz).</summary>
    public string? ChapterQuizId { get; set; }
    /// <summary>Bestehensgrenze in Prozent (0–100).</summary>
    public int PassThresholdPercent { get; set; } = 60;
    /// <summary>Maximale Versuche (0 = unbegrenzt).</summary>
    public int MaxAttempts { get; set; }
}

// ─── Course ──────────────────────────────────────────────────────────────────

[BsonIgnoreExtraElements]
public class Course
{
    [BsonId]
    [BsonRepresentation(BsonType.String)]
    public string Id { get; set; } = Guid.NewGuid().ToString("N");

    public string ElementId { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public Texte Titel { get; set; } = new();
    public string AuthorId { get; set; } = string.Empty;
    public string Status { get; set; } = CourseStatus.Draft;
    public List<Chapter> Chapters { get; set; } = new();
    // F11: Katalog – Filter/Suche
    public List<string> Tags { get; set; } = new();
    public string Level { get; set; } = string.Empty;   // "" | Beginner | Intermediate | Advanced
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public static class CourseStatus
{
    public const string Draft = "Draft";
    public const string Published = "Published";
    public const string Archived = "Archived";
}

public static class CourseLevel
{
    public const string Beginner = "Beginner";
    public const string Intermediate = "Intermediate";
    public const string Advanced = "Advanced";

    public static readonly string[] All = { Beginner, Intermediate, Advanced };
    public static bool IsValid(string? level) => !string.IsNullOrEmpty(level) && All.Contains(level);
}

File: mobile/src/types/course.ts
// ─── Basis & i18n ────────────────────────────────────────────────────────────

export enum Language {
  Russian = 0,
  German = 1,
  English = 2,
}

export interface TextItem {
  text: string;
  language: Language;
}

export interface Texte {
  items: TextItem[];
}

export interface EntityBase {
  elementId: string;
  name: string;
}

// ─── Enums ───────────────────────────────────────────────────────────────────

export enum ChapterContentType {
  Lesson = 0,
  Video = 1,
  Questions = 2,
}

export enum QuestionType {
  OneChoice = 0,
  MultipleChoice = 1,
  OwnAnswer = 2,
  TrueFalse = 3,
  Code = 4,
}

export enum CodeLanguage {
  CSharp = 0,
}

// ─── Quiz & Fragen ───────────────────────────────────────────────────────────

export interface Answer {
  id: string;
  isCorrect: boolean;
  titel: Texte;
  comment: string;
}

export interface Question extends EntityBase {
  titel: Texte;
  questionType: QuestionType;
  answers: Answer[];
  answerValue: string;
  completed: boolean;
  /** Nur bei Code-Fragen gesetzt. SolutionCode + versteckte Testfall-I/O sind für Lerner ausgeblendet. */
  code?: CodeQuestionView;
}

// ─── Code-Aufgaben (F7) ──────────────────────────────────────────────────────

export interface CodeTestCasePreview {
  id: string;
  hidden: boolean;
  /** null für versteckte Testfälle in der Lerner-Sicht. */
  input: string | null;
  expectedOutput: string | null;
}

export interface CodeQuestionView {
  language: CodeLanguage;
  starterCode: string;
  timeLimitMs: number;
  memoryLimitMb: number;
  testCases: CodeTestCasePreview[];
  /** Nur für Autor/Admin gesetzt. */
  solutionCode: string | null;
}

export interface SubmitCodeRequest {
  questionId: string;
  code: string;
}

export interface CodeSubmissionAccepted {
  id: string;
  status: string;
}

export type CodeSubmissionStatus = 'Queued' | 'Running' | 'Completed' | 'Error';

export interface CodeTestCaseResult {
  testCaseId: string;
  hidden: boolean;
  passed: boolean;
  outcome: string;
  durationMs: number;
  input: string | null;
  expectedOutput: string | null;
  actualOutput: string | null;
  stderr: string | null;
}

export interface CodeSubmissionResult {
  id: string;
  questionId: string;
  status: CodeSubmissionStatus;
  outcome: string;
  passedCount: number;
  totalCount: number;
  durationMs: number;
  compileError: string | null;
  errorMessage: string | null;
  testResults: CodeTestCaseResult[];
}

export interface QuestionListModel {
  questions: Question[];
}

// ─── Attempt (POST /api/questions/{id}/attempt) ──────────────────────────────

export interface SubmitAttemptRequest {
  /** OneChoice → single selected answer id */
  answerId?: string;
  /** MultipleChoice → selected answer ids */
  answerIds?: string[];
  /** OwnAnswer → free-text answer */
  textAnswer?: string;
}

/** Server grading result. `answers` carry the revealed isCorrect + explanation comment. */
export interface AttemptResult {
  isCorrect: boolean;
  score: number;
  answers: Answer[];
}

// ─── Content & Kapitel ───────────────────────────────────────────────────────

export interface ChapterContent extends EntityBase {
  courseId: string;
  chapterId: string;
  titel: Texte;
  contentType: ChapterContentType;
  lessonText: string;
  lessonTexte: Texte;
  videoUrl: string;
  questionListId: string;
  sortOrder: number;
  questionLists: QuestionListModel[];
  averageRank: number;
  maxRank: number;
  completed: boolean;
}

export interface Chapter extends EntityBase {
  courseId: string;
  titel: Texte;
  sortOrder: number;
  show: boolean;
  rank: number;
  completed: boolean;
  questions: Question[];
  chapterContent: ChapterContent[];
  // F8: Kapitel-Abschlussquiz
  hasQuiz: boolean;
  passThresholdPercent: number;
  maxAttempts: number;
  quizPassed: boolean;
}

// ─── Kapitel-Abschlussquiz (F8) ──────────────────────────────────────────────

export interface ChapterQuizView {
  chapterId: string;
  passThresholdPercent: number;
  maxAttempts: number;
  questions: Question[];
  attemptsUsed: number;
  bestPercent: number | null;
  passed: boolean;
  attemptsExhausted: boolean;
}

export interface ChapterQuizAnswer {
  questionId: string;
  answerId?: string;
  answerIds?: string[];
}

export interface SubmitChapterQuizRequest {
  answers: ChapterQuizAnswer[];
}

export interface ChapterQuizResult {
  correctCount: number;
  totalCount: number;
  percent: number;
  passed: boolean;
  attemptNo: number;
  maxAttempts: number;
  /** -1 = unbegrenzt. */
  attemptsRemaining: number;
  questions: Question[];
}

// ─── Kurs-Modelle ────────────────────────────────────────────────────────────

export interface Course extends EntityBase {
  titel: Texte;
  chapters: Chapter[];
  // F11: Katalog
  tags: string[];
  level: string; // "" | Beginner | Intermediate | Advanced
}

// ─── API-Response-Modelle ────────────────────────────────────────────────────

export interface ChapterListModel {
  courseId: string;
  courseName: string;
  chapters: Chapter[];
}

export interface ChapterContentListModel {
  courseId: string;
  chapterId: string;
  chapterName: string;
  chapterContent: ChapterContent[];
}
